deep_qa/models/sequence_tagging/verb_semantics_model.py

```diff
@@ -0,0 +1,102 @@
+from deep_qa.training.models import DeepQaModel
+from deep_qa.layers.attention import WeightedSum
+from deep_qa.layers.encoders.bag_of_words import BOWEncoder
+
+
+class VerbSemanticsModel(TextTrainer):
+    """
+    This ``VerbSemanticsModel`` takes as input a sentence and query.
+    Query includes verb and entity spans in the sentence.
+    The model predicts (state-change-type, argument-tags)
+    where, state-change-type: denotes what is happening to the entity, e.g. CREATE/DESTROY/MOVE
+    argument-tags: denotes arguments of the state change, e.g. source and destination location of the entity.
+
+    We use BOW encoder along with a dense layer to predict state-change-type.
+    We use stacked seq2seq encoders followed by a dense layer to predict argument-tags.
+
+    Parameters
+    ----------
+    num_stacked_rnns : int, optional (default: ``1``)
+        The number of ``seq2seq_encoders`` that we should stack on top of each other before
+        predicting tags.
+    instance_type : str
+        Specifies the instance type, currently the only supported type is "VerbSemanticsInstance",
+        which defines things like how the input data is formatted and tokenized.
+    """
+    def __init__(self, params: Params):
+        self.num_stacked_rnns = params.pop('num_stacked_rnns', 1)
+        instance_type = params.pop('instance_type', "VerbSemanticsInstance")
+        self.instance_type = concrete_instances[instance_type]
+        super(VerbSemanticsModel, self).__init__(params)
+
+    @overrides
+    def _instance_type(self):
+        return self.instance_type
+
+    @overrides
+    def _build_model(self):
+        # Input: (sentence, verb-span, entity-span)
+        # Output: (state-change-type, argument-tags)
+
+        # shape: (batch_size, text_length)
+        sentence_input = Input(shape=self._get_sentence_shape(), dtype='int32', name='word_array_input')
+        verb_input = Input(shape=self._get_sentence_shape(), dtype='int32', name='verb_array_input')
+        entity_input = Input(shape=self._get_sentence_shape(), dtype='int32', name='entity_array_input')
+
+        # shape: (batch_size, text_length, embedding_dim)
+        sentence_embedding = self._embed_input(sentence_input)
+        multiply_layer = Multiply()
+        verb_indices = multiply_layer([sentence_input, verb_input])
+        verb_embedding = self._embed_input(verb_indices)
+        entity_indices = multiply_layer([sentence_input, entity_input])
+        entity_embedding = self._embed_input(entity_indices)
+
+        # For state-change-type prediction: We first convert verb and entity into bag-of-words(BOW) representation
+        # and then apply a dense layer with soft-max activation to predict state change type.
+        bow_features = BOWEncoder()
+        weighted_sum_layer = WeightedSum(name="verb_entity_vectors", use_masking=False)
+        verb_entity_vectors = weighted_sum_layer([bow_features(verb_embedding), bow_features(entity_embedding)])
```

Comment:
No - generally Keras handles masking correctly, but there are a few cases with more complex layers which they don't handle. A good rule is that if there is a layer in DeepQA with similar functionality to one that you find in Keras, you should use the DeepQA one. We normally say in the docs as well whether we've overridden it to make it do masking correctly.